### PR TITLE
PHP - Add strong typing and PHPDocs for better dev experience.

### PR DESCRIPTION
**Pull request by Ndiritu**
- Uses strong typing to make it easier to catch bugs in future.
- Adds PHPDocs which enhance dev experience esp for `expand()`
- Fixes minor Exception handling cases.
- Updates PHP to reviewed on README.

closes https://github.com/std-uritemplate/std-uritemplate/issues/54

### CODE REVIEW

File: php/src/StdUriTemplate.php [hunk 1]
```diff
@@ -39,16 +58,18 @@ private static function getMaxChar($buffer, $col) {
             if (empty($value)) {
                 return -1;
             } else {
-                try {
-                    return (int)$value;
-                } catch (NumberFormatException $e) {
```
Comment (Ndiritu) [inline php/src/StdUriTemplate.php:44]: `NumberFormatException` doesn't exist in PHP

File: php/src/StdUriTemplate.php [hunk 2]
```diff
@@ -345,7 +436,7 @@ private static function expandToken($operator, $token, $composite, $maxChar, $fi
             } else {
                 $value = "false";
             }
-        } else if (is_bool($value) || is_int($value) || is_float($value) || is_double($value)) {
```
Comment (Ndiritu) [inline php/src/StdUriTemplate.php:348]: `is_bool` is already checked in previous condition

File: php/src/StdUriTemplate.php (hunk 1 above)
Comment (andreaTP) [reply php/src/StdUriTemplate.php:44]: To keep the implementation close to the others, we should verify and eventually catch the exception here, we can use an additional check like `is_numeric` to do the right validation.

File: php/src/StdUriTemplate.php [hunk 3]
```diff
@@ -82,7 +109,11 @@ private static function expandImpl($str, $substitutions) {
                     break;
                 case '}':
                     if ($token !== null) {
-                        $expanded = self::expandToken($operator, $token, $composite, self::getMaxChar($maxCharBuffer, $i), $firstToken, $substitutions, $result, $i);
+                        try {
+                            $expanded = self::expandToken($operator, $token, $composite, self::getMaxChar($maxCharBuffer), $firstToken, $substitutions, $result, $i);
+                        } catch (TypeError $ex) {
```
Comment (andreaTP) [inline php/src/StdUriTemplate.php:114]: Here we are catching ways more than exceptions coming from `getMaxChar`, I suggest we revert this detail.

File: php/src/StdUriTemplate.php [hunk 4]
```diff
@@ -91,12 +122,16 @@ private static function expandImpl($str, $substitutions) {
...
+                        throw new InvalidArgumentException("Failed to expand token, invalid at col: $i");
                     }
                     break;
                 case ',':
                     if ($token !== null) {
-                        $expanded = self::expandToken($operator, $token, $composite, self::getMaxChar($maxCharBuffer, $i), $firstToken, $substitutions, $result, $i);
+                        try {
+                            $expanded = self::expandToken($operator, $token, $composite, self::getMaxChar($maxCharBuffer), $firstToken, $substitutions, $result, $i);
+                        } catch (TypeError $ex) {
+                            throw new InvalidArgumentException("Cannot parse max chars at col: $i");
```
Comment (andreaTP) [inline php/src/StdUriTemplate.php:133]: as above, this check should not be propagated.

File: php/src/StdUriTemplate.php [hunk 5]
```diff
@@ -304,38 +375,58 @@ private static function isMap($value) {
...
+     * @return bool
+     */
+    private static function isEmpty(string $substType, $value): bool {
         if ($value === null) {
             return true;
         } else {
             switch ($substType) {
                 case 'STRING':
                     return false;
-                case 'LIST':
```
Comment (andreaTP) [inline php/src/StdUriTemplate.php:326]: Can you please leave `LIST` before `MAP`?
The check to determine if it's a List is much much cheaper in terms of performance.

File: php/src/StdUriTemplate.php (hunk 2 above)
Comment (andreaTP) [reply php/src/StdUriTemplate.php:348]: correct 👍 thanks!

**Review by andreaTP — CHANGES_REQUESTED** [review]
@Ndiritu thanks a lot for those improvements!
I really love them as they improve the quality of this PHP implementation.

Added a few comments inline but the only important one is to restore the "check + exception throw" in the `getMaxChar` method, please 🙏  I'm convinced that the current approach is more aligned with other implementations and easier to debug.

File: php/src/StdUriTemplate.php (hunk 3 above)
Comment (Ndiritu) [reply php/src/StdUriTemplate.php:114]: Yeah, fair point.

File: php/src/StdUriTemplate.php (hunk 1 above)
Comment (Ndiritu) [reply php/src/StdUriTemplate.php:44]: Used `is_int` to validate the type as well.

File: php/src/StdUriTemplate.php (hunk 5 above)
Comment (Ndiritu) [reply php/src/StdUriTemplate.php:326]: Added.

**Review by andreaTP — APPROVED** [review]
lgtm, great work @Ndiritu , thanks a lot!

**Pull request merged**